agent/internal/containers.go

```diff
@@ -144,32 +145,45 @@ func (c *containerManager) overwriteSpec(
 		slotIds = append(slotIds, strconv.Itoa(d.ID))
 	}
 	spec.RunSpec.ContainerConfig.Labels[dockerContainerDevicesLabel] = strings.Join(slotIds, ",")
-	if len(cont.Devices) > 0 && cont.Devices[0].Type == device.GPU {
-		spec.RunSpec.HostConfig.Runtime = nvidiaRuntime
+
+	spec = c.overwriteSpecforGPUs(cont, spec)
+
+	return spec
+}
+
+func (c *containerManager) overwriteSpecforGPUs(
+	cont cproto.Container, spec container.Spec,
+) container.Spec {
+	var gpuUUIDs []string
+	for _, d := range cont.Devices {
+		if d.Type == device.GPU {
+			gpuUUIDs = append(gpuUUIDs, d.UUID)
+		}
+	}
+	useGPU := len(gpuUUIDs) > 0
+	spec.RunSpec.ContainerConfig.Env = append(
+		spec.RunSpec.ContainerConfig.Env,
+		fmt.Sprintf("DET_USE_GPU=%t", useGPU),
+	)
```

Comment:
blocking: I think not having this environment variable set with the rest of the environment variables can get us a pretty subtle bug in the future.

I think we can move some of this logic into a function and use it in both `containerEnvVars` and this function. While we would loop through the devices twice, the slice will be ~10 so it should be fine. Could avoid the double loop by passing in more arguments if we think that's the right move

Reply:
Hmm, what's the scenario you're imagining that would lead to a bug? I don't think I see it.

Reply:
Before the logic here is:
```
<SET GLOBAL ENV VARS>
<SET CONTAINER ENV VARS>
<SET A BUNCH OF LABELS>
```

What is proposed here is:
```
<SET GLOBAL ENV VARS>
<SET CONTAINER ENV VARS>
<SET A BUNCH OF LABELS>
<SET A FEW OTHER THINGS, INCLUDING AN ENV VAR>
```

For now this works functionally speaking. However reading through it, we are now setting these container env vars in multiple places. This makes it harder to find "what are the environment varialbes we are setting". Especially if these "other things" start growing more. And if we miss an important env var, that would be a bug that would be a bit subtle. What I'm proposing is making the workflow look more like:

```
<SET GLOBAL ENV VARS>
<SET CONTAINER ENV VARS (including DET_USE_GPU)>
<SET A BUNCH OF LABELS>
<SET DEVICE REQUESTS>
```

This maintains the clean separation of overwriting different parts of the configs and ensures it's very transparent to the reader what is being overwritten where. If you look at the current logic it is is not objective-based (e.g. "overwrite for GPUs") but based on the part of the spec that is being overwritten. I believe we should maintain that pattern and not break it.

Reply:
:+1: